### DISCUSSION TITLE
DHCP Based IP Alloc

### DISCUSSION THREAD

**Discussion by sudhikan** [magma/magma#5235] [Show and tell]
cc: Peter Kapastos 

Here is my setup for DHCP based allocation: 

1. eth0 - set in /etc/network/interfaces.d/eth0 as 
```
auto eth0
  iface eth0 inet dhcp
```
During runtime, you'll notice that the IP address is removed from the eth0 interface and assigned to the `uplink_br0` bridge instead. 

2. Make sure eth0 is on the same broadcast domain as a DHCP server (or the relay agent... depending on your setup).

3. In the NMS make sure IP Allocation method is set to "DHCP_Broadcast"
![image](https://user-images.githubusercontent.com/38921899/109577584-af5bf380-7aaa-11eb-97e2-def0fb1d38b7.png)

You're good to go after this. 

In my setup, I made the DHCP server send the gateway address as the SVI interface on my nexus L3 device.

AGW will respond with the mac address of the gtp_br0 interface AFAIK for ARP. 

Mobilityd showing the IP address of the attached UE and the gtp_br0 IP addr. 
![image](https://user-images.githubusercontent.com/38921899/109574337-96e8da80-7aa4-11eb-8d6c-07d0f9aa0d21.png)

Pinging the Attached UE from the upstream DHCP server and showing the arp resolution. 
![image](https://user-images.githubusercontent.com/38921899/109574408-b08a2200-7aa4-11eb-907d-d078bc75440e.png)

**Comment by pkapatsos-blinq** [comment]
Hi so this is for single apn and the ethernet Interface is a Access-port correct ?  I can get this but no response ,
![image](https://user-images.githubusercontent.com/61711855/109575529-fd2f2680-7abf-11eb-9da7-95856fbef6eb.png)
![image](https://user-images.githubusercontent.com/61711855/109575637-3d8ea480-7ac0-11eb-8164-997a79f683ef.png)

Note I have DHCP Broadcast as the setting , only way I can get the IP to the CPE  form the DCHP server when I NAT it uses the ip pool form the AGW custom settings so NAT .
![image](https://user-images.githubusercontent.com/61711855/109575906-bbeb4680-7ac0-11eb-8140-5b5df86da279.png)


We really need multiple APNs working , I think both issues are the same though for me single APN gets IP no data passes multi APN I can get IPs to the CPE   .getting late here so can we continue tomorrow please think we re 3 hours dif , 
let me know what I can show you vie tcpdump or pcacp debug logs etc. thanks

**Reply by sudhikan** [reply]
I'm so sorry.. I pasted the wrong screenshot! It should be DHCP_Broadcast. My setup was DHCP_Broadcast but got the screenshot from another setup :/ 

Yes, the interface is an access port on a vlan. Can you disable "Static IP Assignments"? AFAIK this is used for static IP assignment if IP per APN is configured in the subscriber information (cc: @pshelar) and not for DHCP_Broadcast.

**Reply by ulaskozat** [reply]
@pkapatsos-blinq  besides disabling "Static IP Assignments", are you configuring the vlan tags for APNs?

**Reply by pkapatsos-blinq** [reply]
Yes @ulaskozat with all as dynamic not working with TRUNK port and vlan tags on APNs , 
![image](https://user-images.githubusercontent.com/61711855/110037490-791dae80-7d0c-11eb-8c14-00515b9c63bb.png)

We needed a config asap to manage CPEs for a friendly so will add here on this issue, thought it was that I was tagging eth0 and had IPs but even removing just now looking at it again .Will update

However with @pshelar help we have a configuration getting base APN via DHCP via upstream cisco switch as the server and port as Access , the 2nd apn as static and its subnet as secondary IP on same VLAN.. at least this allows for use to move forward a bit and be able to manage CPEs, Also note had to apply this to work correctly https://github.com/magma/magma/pull/4338 . However still minor issue with ARPs , they seem to not get updated and must flush the table on the switch router after reconnection. 
Why is ip allocation a network wide function ahh , any way around it is that you have to group AGW with what method you are using by creating new network .
We need a feature that the AGW provides the IP POOLs via APN/PDNs on the AGW  Like you do for NAT as well hoping that's coming soon , OAI code has this under SPGW.
For any one else to use DUAL APN via one dynamic and one static no trunk port  here you go,
set AGW to NAT or custom via NMS , 
![image](https://user-images.githubusercontent.com/61711855/110035631-ec71f100-7d09-11eb-9ab3-2e5c64474bf4.png)

then under network in EPC tab select as DHCP allocation and set .push to AGW. wait for mobilityd to restart . 
![image](https://user-images.githubusercontent.com/61711855/110035523-cc423200-7d09-11eb-83df-3828e557b589.png)
Make sure that the uplink_br0 has the IP from eth0 ( this only works if eth0 is dhcp not static)
uplink_br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 00:90:0b:62:fa:20 brd ff:ff:ff:ff:ff:ff
    inet 192.168.39.29/24 brd 192.168.39.255 scope global uplink_br0
       valid_lft forever preferred_lft forever

Then set the Subscriber to get a static for second apn, 
![image](https://user-images.githubusercontent.com/61711855/110036017-63a78500-7d0a-11eb-8785-969f2cdf2f4b.png)

outside switch /rotuer config 
![image](https://user-images.githubusercontent.com/61711855/110038060-3e684600-7d0d-11eb-9733-27705b7f3a90.png)

**Comment by jpad-freedomfi** [comment]
@pkapatsos-blinq above you said:
`We need a feature that the AGW provides the IP POOLs via APN/PDNs on the AGW Like you do for NAT as well hoping that's coming soon , OAI code has this under SPGW.`

We have just started looking in to the same. Agree with the concept there.
